File: lib/py/network_flow.py
from collections import deque

class Dinic:
    def __init__(self, n):
        self.n = n
        self.adj = [[] for _ in range(n + 1)]
        self.level = []
        
    def add_edge(self, u, v, cap):
        # [to, cap, flow, rev_index]
        self.adj[u].append([v, cap, 0, len(self.adj[v])])
        self.adj[v].append([u, 0, 0, len(self.adj[u]) - 1])
        
    def bfs(self, s, t):
        self.level = [-1] * (self.n + 1)
        self.level[s] = 0
        q = deque([s])
        while q:
            cur = q.popleft()
            for v, cap, flow, rev in self.adj[cur]:
                if cap - flow > 0 and self.level[v] == -1:
                    self.level[v] = self.level[cur] + 1
                    q.append(v)
        return self.level[t] != -1
        
    def dfs(self, cur, t, pushed, ptr):
        if pushed == 0 or cur == t:
            return pushed
            
        for i in range(ptr[cur], len(self.adj[cur])):
            ptr[cur] = i
            v, cap, flow, rev = self.adj[cur][i]
            
            if self.level[v] != self.level[cur] + 1 or cap - flow == 0:
                continue
                
            tr = self.dfs(v, t, min(pushed, cap - flow), ptr)
            if tr == 0: continue
            
            self.adj[cur][i][2] += tr # flow update
            self.adj[v][rev][2] -= tr # rev flow update
            return tr
            
        return 0
        
    def max_flow(self, s, t):
        flow = 0
        while self.bfs(s, t):
            ptr = [0] * (self.n + 1)
            while True:
                pushed = self.dfs(s, t, float('inf'), ptr)
                if pushed == 0: break
                flow += pushed
        return flow

File: lib/cpp/network_flow.cpp
#include <algorithm>
#include <queue>
#include <vector>


using namespace std;

const int INF = 1e9;

struct Edge {
  int to, cap, flow, rev;
};

vector<vector<Edge>> adj;
vector<int> level, ptr;

void add_edge(int from, int to, int cap) {
  adj[from].push_back({to, cap, 0, (int)adj[to].size()});
  adj[to].push_back({from, 0, 0, (int)adj[from].size() - 1});
}

// BFS: 레벨 그래프 생성
bool bfs(int s, int t) {
  fill(level.begin(), level.end(), -1);
  level[s] = 0;
  queue<int> q;
  q.push(s);
  while (!q.empty()) {
    int cur = q.front();
    q.pop();
    for (const auto &edge : adj[cur]) {
      if (edge.cap - edge.flow > 0 && level[edge.to] == -1) {
        level[edge.to] = level[cur] + 1;
        q.push(edge.to);
      }
    }
  }
  return level[t] != -1;
}

// DFS: 증가 경로 찾기
int dfs(int cur, int t, int pushed) {
  if (pushed == 0 || cur == t)
    return pushed;
  for (int &cid = ptr[cur]; cid < adj[cur].size(); cid++) {
    auto &edge = adj[cur][cid];
    if (level[cur] + 1 != level[edge.to] || edge.cap - edge.flow == 0)
      continue;
    int tr = dfs(edge.to, t, min(pushed, edge.cap - edge.flow));
    if (tr == 0)
      continue;
    edge.flow += tr;
    adj[edge.to][edge.rev].flow -= tr;
    return tr;
  }
  return 0;
}

// Dinic Algorithm
int dinic(int s, int t) {
  int flow = 0;
  while (bfs(s, t)) {
    fill(ptr.begin(), ptr.end(), 0);
    while (int pushed = dfs(s, t, INF)) {
      flow += pushed;
    }
  }
  return flow;
}

// 초기화: adj.assign(n + 1, vector<Edge>()); level.resize(n + 1); ptr.resize(n
// + 1);
